File: Sources/ExcalidrawMath/Range.swift
import Foundation

/// An inclusive numeric range `[start, end]` (`packages/math/src/range.ts`).
public struct InclusiveRange: Equatable, Sendable {
    public var start: Double
    public var end: Double

    public init(_ start: Double, _ end: Double) {
        self.start = start
        self.end = end
    }

    /// Whether this range overlaps `other` (`rangesOverlap`).
    public func overlaps(_ other: InclusiveRange) -> Bool {
        if start <= other.start { return end >= other.start }
        return other.end >= start
    }

    /// The overlapping range, or `nil` when disjoint (`rangeIntersection`).
    public func intersection(_ other: InclusiveRange) -> InclusiveRange? {
        let lo = Swift.max(start, other.start)
        let hi = Swift.min(end, other.end)
        return lo <= hi ? InclusiveRange(lo, hi) : nil
    }

    /// Whether `value` lies inside the range (`rangeIncludesValue`).
    public func includes(_ value: Double) -> Bool {
        value >= start && value <= end
    }
}

File: web/packages/excalidraw-svelte/src/math/range.ts
/** An inclusive numeric range `[start, end]`. (parity: Range.swift) */
export class InclusiveRange {
  constructor(
    public start: number,
    public end: number,
  ) {}

  /** Whether this range overlaps `other` (`rangesOverlap`). */
  overlaps(other: InclusiveRange): boolean {
    if (this.start <= other.start) return this.end >= other.start;
    return other.end >= this.start;
  }

  /** The overlapping range, or `null` when disjoint (`rangeIntersection`). */
  intersection(other: InclusiveRange): InclusiveRange | null {
    const lo = Math.max(this.start, other.start);
    const hi = Math.min(this.end, other.end);
    return lo <= hi ? new InclusiveRange(lo, hi) : null;
  }

  /** Whether `value` lies inside the range (`rangeIncludesValue`). */
  includes(value: number): boolean {
    return value >= this.start && value <= this.end;
  }
}
